File: java/src/main/java/com/danifoldi/dml/type/DmlObject.java
package com.danifoldi.dml.type;

import com.danifoldi.dml.DmlSerializer;

import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public class DmlObject extends DmlCommentableValue implements DmlDocument {
    private Map<DmlKey, DmlValue> value;

    public DmlObject(Map<DmlKey, DmlValue> value) {
        this.value = value;
    }

    public void set(DmlKey key, DmlValue value) {
        this.value.put(key, value);
    }

    public void set(String key, DmlValue value) {
        this.value.put(new DmlKey(key), value);
    }

    public DmlValue get(DmlKey key) {
        return this.value.get(key);
    }

    public DmlArray getArray(DmlKey key) {
        return this.value.get(key).asArray();
    }

    public DmlBoolean getBoolean(DmlKey key) {
        return this.value.get(key).asBoolean();
    }

    public DmlNumber getNumber(DmlKey key) {
        return this.value.get(key).asNumber();
    }

    public DmlObject getObject(DmlKey key) {
        return this.value.get(key).asObject();
    }

    public DmlString getString(DmlKey key) {
        return this.value.get(key).asString();
    }

    public Set<DmlKey> keys() {
        return this.value.keySet();
    }

    public DmlValue get(String key) {
        return this.value.get(new DmlKey(key));
    }

    public DmlArray getArray(String key) {
        return this.value.get(new DmlKey(key)).asArray();
    }

    public DmlBoolean getBoolean(String key) {
        return this.value.get(new DmlKey(key)).asBoolean();
    }

    public DmlNumber getNumber(String key) {
        return this.value.get(new DmlKey(key)).asNumber();
    }

    public DmlObject getObject(String key) {
        return this.value.get(new DmlKey(key)).asObject();
    }

    public DmlString getString(String key) {
        return this.value.get(new DmlKey(key)).asString();
    }

    @Override
    public String serialize(int indent) {
        String dense = value.keySet().stream().map(v -> v.serialize(2) + ": " + value.get(v).serialize(0)).collect(Collectors.joining(", "));
        return DmlSerializer.serializeComment(comment(), indent) + (!dense.contains("\n") && dense.length() <= 80 ? (value.isEmpty() ? "{}" : "{ " + dense + " }") : DmlSerializer.indent("{\n" + value.keySet().stream().sorted().map(v -> v.serialize(2) + ": " + value.get(v).serialize(2)).collect(Collectors.joining("\n")) + "\n}", indent));
    }

    @Override
    public String serializeDocument() {
        return DmlSerializer.serializeComment(comment(), 0) + DmlSerializer.indent(value.keySet().stream().map(v -> v.serialize(0) + ": " + value.get(v).serialize(0)).collect(Collectors.joining("\n")), 0);
    }

    @Override
    public String toString() {
        return "DmlObject{" + value + "}";
    }
}

File: javascript/src/types/DmlObject.js
import DmlKey from './DmlKey.js'
import DmlSerializer from '../DmlSerializer.js'
import DmlCommentableValue from './DmlCommentableValue.js'

class DmlObject extends DmlCommentableValue {
  #map

  constructor(map) {
    super()
    this.#map = map
  }

  set(key, value) {
    if (!(key instanceof DmlKey)) {
      key = new DmlKey(key)
    }
    this.#map.set(key.value(), [key, value])
  }

  get(key) {
    if (key instanceof DmlKey) {
      key = key.value()
    }
    return this.#map.get(key)[1]
  }

  serialize(indent) {
    let dense = [...this.#map.keys()]
      .map(v => this.#map.get(v))
      .map(v => v[0].serialize(2) + ': ' + v[1].serialize(0)).join(', ')
    return DmlSerializer.serializeComment(this.comment(), indent)
      + (dense.indexOf('\n') === -1 && dense.length <= 80)
      ? (this.#map.size === 0 ? '{}' : '{' + dense + '}')
      : DmlSerializer.indent(
        '{\n'
        + [...this.#map.keys()].map(v => this.#map.get(v))
          .map(v => v[0].serialize(2) + ': ' + v[1].serialize(0)).join('\n')
        + '\n}',
        indent
      )
  }

  serializeDocument() {
    return DmlSerializer.serializeComment(this.comment(), 0)
      + DmlSerializer.indent(
        [...this.#map.keys()]
          .map(v => this.#map.get(v))
          .map(v => v[0].serialize(0) + ': ' + v[1].serialize(0)).join(',\n'),
        0
      )
  }

  get [Symbol.toStringTag]() {
    return `DmlObject{${[...this.#map.values()].map(v => v[0] + ': ' + v[1]).join(', ')}}`
  }
}

export default DmlObject
